packages/runtime/src/runtime/hooks/useEnsembleAction.tsx

```diff
@@ -299,6 +315,179 @@ export const useInvokeAPI: EnsembleActionHook<InvokeAPIAction> = (action) => {
   return invokeApi;
 };
 
+export const useConnectSocket: EnsembleActionHook<ConnectSocketAction> = (
+  action,
+) => {
+  const { sockets, setData } = useScreenData();
+  const [isComplete, setIsComplete] = useState<boolean>();
+  const [isLoading, setIsLoading] = useState<boolean>();
+
+  const socket = useMemo(
+    () => sockets?.find((model) => model.name === action?.name),
+    [action, sockets],
+  );
+
+  const onSocketConnectAction = useEnsembleAction(socket?.onSuccess);
+  const onMessageReceiveAction = useEnsembleAction(socket?.onReceive);
+  const onSocketDisconnectAction = useEnsembleAction(socket?.onDisconnect);
+
+  const connectSocket = useMemo(() => {
+    if (!socket) {
+      return;
+    }
+
+    const callback = (): void => {
+      setIsComplete(false);
+    };
+    return { callback };
+  }, [socket]);
+
+  useEffect(() => {
+    if (!socket || isComplete !== false || isLoading) {
+      return;
+    }
+
+    const fireRequest = (): void => {
+      setIsLoading(true);
+      try {
+        const ws = new WebSocket(socket.uri);
+
+        ws.onopen = (): void => {
+          onSocketConnectAction?.callback();
+        };
+
+        ws.onmessage = (e: MessageEvent): void => {
+          onMessageReceiveAction?.callback({ data: e.data as unknown });
+        };
+
+        ws.onclose = (): void => {
+          onSocketDisconnectAction?.callback();
+        };
+
+        setData(socket.name, ws);
+      } catch (e) {
+        logError(e);
+      } finally {
+        setIsLoading(false);
+        setIsComplete(true);
+      }
+    };
+
+    fireRequest();
+  }, [
+    socket,
+    isComplete,
+    isLoading,
+    setData,
+    onSocketConnectAction,
+    onMessageReceiveAction,
+    onSocketDisconnectAction,
+  ]);
+
+  return connectSocket;
+};
+
+export const useMessageSocket: EnsembleActionHook<SendSocketMessageAction> = (
+  action,
+) => {
+  const { sockets, data } = useScreenData();
+  const [isComplete, setIsComplete] = useState<boolean>();
+  const [isLoading, setIsLoading] = useState<boolean>();
+  const [context, setContext] = useState<{ [key: string]: unknown }>();
+  const evaluatedInputs = useEvaluate(action?.message, { context });
+
+  const socket = useMemo(
+    () => sockets?.find((model) => model.name === action?.name),
+    [action, sockets],
+  );
+
+  const sendSocketMessage = useMemo(() => {
+    if (!socket) {
+      return;
+    }
+
+    const callback = (args: unknown): void => {
+      setIsComplete(false);
+      setContext(args as { [key: string]: unknown });
+    };
+    return { callback };
+  }, [socket]);
+
+  useEffect(() => {
+    if (!socket || isComplete !== false || isLoading) {
+      return;
+    }
+
+    const fireRequest = (): void => {
+      setIsLoading(true);
+      try {
+        const socketInstance = data[socket.name] as WebSocket;
+        if (socketInstance) {
+          socketInstance.send(JSON.stringify(evaluatedInputs));
```

Comment:
Are these awaitable?

Reply:
No, because this method add the message in queue of the data need to be transmitted on server through socket. so this will just add the message in queue and return response.